File: ml_service/predict.py
"""
PriceWatch AI — Price Prediction Service
==========================================
Simple but accurate ML predictor using Linear + Polynomial Regression.
No complex setup needed — pure Python, runs on Render for free.

How it works (for your viva):
  1. Fetch price history from Supabase PostgreSQL
  2. Convert dates → day numbers (day 0, 1, 2 ...)
  3. Fit a polynomial regression curve through the price points
  4. Extend the curve 7 days into the future
  5. Analyze the slope to generate buying insights
  6. Return JSON with predictions + insight + confidence score

Start as HTTP server:  python predict.py --serve
Test one product:      python predict.py --product_id 1
"""

import os, sys, json, argparse
from datetime import datetime, timedelta, timezone

# ── Check required libraries ──────────────────────────────────
try:
    import psycopg2
    from psycopg2.extras import RealDictCursor
except ImportError:
    print(json.dumps({"error": "Run: pip install psycopg2-binary"}))
    sys.exit(1)

try:
    import numpy as np
    from sklearn.linear_model import LinearRegression
    from sklearn.preprocessing import PolynomialFeatures
    from sklearn.metrics import r2_score
except ImportError:
    print(json.dumps({"error": "Run: pip install numpy scikit-learn"}))
    sys.exit(1)

try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass

# ─────────────────────────────────────────────────────────────
# DATABASE
# ─────────────────────────────────────────────────────────────
def get_conn():
    url = os.environ.get("DATABASE_URL", "")
    if not url:
        raise Exception("DATABASE_URL not set in environment.")
    return psycopg2.connect(url, sslmode='require')

def fetch_history(product_id):
    sql = """
        SELECT price::float AS price, recorded_at
        FROM   price_history
        WHERE  product_id = %s
        ORDER  BY recorded_at ASC
    """
    conn = get_conn()
    try:
        with conn.cursor(cursor_factory=RealDictCursor) as cur:
            cur.execute(sql, (product_id,))
            return [dict(r) for r in cur.fetchall()]
    finally:
        conn.close()

def fetch_product(product_id):
    conn = get_conn()
    try:
        with conn.cursor(cursor_factory=RealDictCursor) as cur:
            cur.execute("SELECT name, category, url FROM products WHERE product_id = %s", (product_id,))
            row = cur.fetchone()
            return dict(row) if row else {}
    finally:
        conn.close()

# ─────────────────────────────────────────────────────────────
# ML PIPELINE
# ─────────────────────────────────────────────────────────────
def to_day_numbers(history):
    """Convert timestamps to numeric day offsets from first entry."""
    dates, prices = [], []
    for row in history:
        dt = row['recorded_at']
        if isinstance(dt, str):
            dt = datetime.fromisoformat(dt.replace('Z','+00:00'))
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        dates.append(dt)
        prices.append(float(row['price']))
    base = dates[0]
    days = [(d - base).total_seconds() / 86400 for d in dates]
    return np.array(days).reshape(-1,1), np.array(prices), base, dates

def train(X, y):
    """
    Polynomial Regression (degree 2).
    degree=1 for <5 points (linear), degree=2 otherwise (curve).
    R² score tells us how well the model fits (0=bad, 1=perfect).
    """
    degree = 1 if len(X) < 5 else 2
    poly   = PolynomialFeatures(degree=degree, include_bias=False)
    Xp     = poly.fit_transform(X)
    model  = LinearRegression().fit(Xp, y)
    r2     = max(0.0, r2_score(y, model.predict(Xp)))
    return model, poly, r2

def predict_next_7(model, poly, X, base_date):
    """Extend the regression curve 7 days beyond the last data point."""
    last_day = float(X[-1][0])
    results  = []
    for i in range(1, 8):
        day        = last_day + i
        pred_price = max(1.0, model.predict(poly.transform([[day]]))[0])
        future_dt  = base_date + timedelta(days=day)
        results.append({
            "date":  future_dt.strftime("%Y-%m-%d"),
            "price": round(float(pred_price), 2),
        })
    return results

def make_insight(current_price, preds, y, r2):
    """Generate human-readable insight from predictions."""
    prices      = [p['price'] for p in preds]
    min_price   = min(prices)
    max_price   = max(prices)
    min_idx     = prices.index(min_price)
    best_day    = preds[min_idx]['date']
    last_pred   = prices[-1]
    pct_change  = (last_pred - current_price) / current_price * 100

    # Detect sudden single-day drops > 5%
    drop_detected = any(
        (prices[i] - prices[i-1]) / prices[i-1] * 100 < -5
        for i in range(1, len(prices))
    )

    # Trend classification
    if pct_change < -3:
        trend = 'falling'
    elif pct_change > 3:
        trend = 'rising'
    else:
        trend = 'stable'

    # Confidence score: R² × data-quantity factor (needs 10+ points for 100%)
    data_factor = min(1.0, len(y) / 10)
    confidence  = round(max(0, min(100, r2 * data_factor * 100)), 1)

    savings = round(current_price - min_price, 2)

    if trend == 'falling':
        if drop_detected:
            insight = (f"🚨 Sharp drop detected! Price expected to fall by ₹{abs(savings):,.0f} "
                       f"in the next {min_idx+1} day(s). Best time to buy: {best_day}.")
        else:
            insight = (f"📉 Price trending down by {abs(pct_change):.1f}% over 7 days. "
                       f"Consider waiting — best price expected on {best_day} (₹{min_price:,.0f}).")
    elif trend == 'rising':
        insight = (f"📈 Price likely to rise by {abs(pct_change):.1f}% over 7 days. "
                   f"Buy now before it increases. Current price is the best available.")
    else:
        insight = (f"➡️ Price is stable (±{abs(pct_change):.1f}%). "
                   f"No rush — lowest predicted price: ₹{min_price:,.0f} on {best_day}.")

    if confidence < 40:
        insight += f" ⚠️ Low confidence ({len(y)} data points — more price updates improve accuracy)."

    return {
        "insight":         insight,
        "trend":           trend,
        "confidence":      confidence,
        "drop_detected":   drop_detected,
        "best_buy_day":    best_day,
        "min_predicted":   round(min_price, 2),
        "max_predicted":   round(max_price, 2),
        "expected_change_pct": round(pct_change, 2),
    }

# ─────────────────────────────────────────────────────────────
# MAIN PREDICT FUNCTION
# ─────────────────────────────────────────────────────────────
def predict(product_id):
    """Full pipeline: fetch → train → predict → insight → return JSON."""
    try:
        history = fetch_history(product_id)
        product = fetch_product(product_id)

        if len(history) < 3:
            return {
                "product_id": product_id,
                "product_name": product.get('name',''),
                "error": f"Need at least 3 price records. Currently have {len(history)}.",
                "prediction": [],
                "insight": "Add more price data to enable AI predictions.",
                "confidence": 0,
                "trend": "unknown",
            }

        X, y, base_date, dates = to_day_numbers(history)
        model, poly, r2        = train(X, y)
        predictions            = predict_next_7(model, poly, X, base_date)

        current_price = float(y[-1])
        analysis      = make_insight(current_price, predictions, y, r2)

        # Last 14 points for history graph
        history_graph = []
        for row in history[-14:]:
            dt = row['recorded_at']
            history_graph.append({
                "date":  dt.strftime("%Y-%m-%d") if hasattr(dt,'strftime') else str(dt)[:10],
                "price": float(row['price']),
            })

        return {
            "product_id":      product_id,
            "product_name":    product.get('name',''),
            "current_price":   current_price,
            "data_points":     len(history),
            "r2_score":        round(r2, 4),
            "prediction":      predictions,
            "history":         history_graph,
            **analysis,
        }

    except Exception as e:
        return {
            "product_id": product_id,
            "error": str(e),
            "prediction": [],
            "insight": "Prediction failed.",
            "confidence": 0,
            "trend": "unknown",
        }

# ─────────────────────────────────────────────────────────────
# FLASK HTTP SERVER  (Node.js calls this)
# ─────────────────────────────────────────────────────────────
def run_server():
    try:
        from flask import Flask, jsonify
        from flask_cors import CORS
    except ImportError:
        print("Run: pip install flask flask-cors")
        sys.exit(1)

    app = Flask(__name__)
    CORS(app)

    @app.route('/predict/<int:product_id>')
    def predict_route(product_id):
        return jsonify(predict(product_id))

    @app.route('/health')
    def health():
        return jsonify({"status": "ok", "service": "PriceWatch AI"})

    port = int(os.environ.get("ML_PORT", 5001))
    print(f"\n🤖 PriceWatch AI running → http://localhost:{port}")
    app.run(host='0.0.0.0', port=port, debug=False)

# ─────────────────────────────────────────────────────────────
# CLI
# ─────────────────────────────────────────────────────────────
if __name__ == "__main__":
    ap = argparse.ArgumentParser()
    ap.add_argument('--product_id', type=int)
    ap.add_argument('--serve', action='store_true')
    args = ap.parse_args()

    if args.serve:
        run_server()
    elif args.product_id:
        print(json.dumps(predict(args.product_id), indent=2, default=str))
    else:
        ap.print_help()

File: backend/routes/predict.js
/**
 * GET /api/predict/:product_id
 * Pure JS Linear Regression predictor — no Python needed.
 * Works on Render free tier without any extra service.
 * Cached 30 minutes to avoid repeated computation.
 */
const express         = require('express');
const db              = require('../db');
const { requireAuth } = require('../middleware/auth');
const router          = express.Router();

// 30-minute cache
const CACHE     = new Map();
const CACHE_TTL = 30 * 60 * 1000;

router.get('/:product_id', requireAuth, async (req, res) => {
  const pid = parseInt(req.params.product_id);
  if (isNaN(pid)) return res.status(400).json({ error: 'Invalid product ID.' });

  // Cache hit
  const hit = CACHE.get(pid);
  if (hit && Date.now() - hit.ts < CACHE_TTL)
    return res.json({ ...hit.data, cached: true });

  // Verify ownership
  const [prod] = await db.query(
    'SELECT product_id, name, category FROM products WHERE product_id = ? AND added_by = ?',
    [pid, req.session.user.user_id]
  ).catch(() => [[]]);
  if (!prod.length) return res.status(404).json({ error: 'Product not found.' });

  // Fetch price history
  const [history] = await db.query(
    `SELECT price::float AS price, recorded_at
     FROM price_history WHERE product_id = ?
     ORDER BY recorded_at ASC`,
    [pid]
  ).catch(() => [[]]);

  if (history.length < 3) {
    return res.json({
      product_id:   pid,
      product_name: prod[0].name,
      error:        `Need at least 3 price records (currently ${history.length}). Update price ${3 - history.length} more time(s).`,
      prediction:   [],
      insight:      'Not enough data yet.',
      trend:        'unknown',
      confidence:   0,
    });
  }

  const result = buildPrediction(pid, prod[0].name, prod[0].category, history);
  CACHE.set(pid, { data: result, ts: Date.now() });
  res.json({ ...result, cached: false });
});

// ── Core prediction engine ─────────────────────────────────────
function buildPrediction(pid, name, category, history) {
  const prices = history.map(r => parseFloat(r.price));
  const n      = prices.length;

  // --- Linear regression: y = m*x + b ---
  const xs   = Array.from({ length: n }, (_, i) => i);
  const xBar = xs.reduce((a, b) => a + b, 0) / n;
  const yBar = prices.reduce((a, b) => a + b, 0) / n;
  const num  = xs.reduce((s, x, i) => s + (x - xBar) * (prices[i] - yBar), 0);
  const den  = xs.reduce((s, x)    => s + (x - xBar) ** 2, 0);
  const m    = den !== 0 ? num / den : 0;
  const b    = yBar - m * xBar;

  // R² score (model accuracy)
  const ssTot = prices.reduce((s, p) => s + (p - yBar) ** 2, 0);
  const ssRes = prices.reduce((s, p, i) => s + (p - (m * i + b)) ** 2, 0);
  const r2    = ssTot > 0 ? Math.max(0, 1 - ssRes / ssTot) : 0;

  // --- Predict next 7 days ---
  const lastDate   = new Date(history[history.length - 1].recorded_at);
  const prediction = [];
  for (let i = 1; i <= 7; i++) {
    const futureDate = new Date(lastDate);
    futureDate.setDate(futureDate.getDate() + i);
    const predPrice = Math.max(1, m * (n - 1 + i) + b);
    prediction.push({
      date:  futureDate.toISOString().split('T')[0],
      price: Math.round(predPrice * 100) / 100,
    });
  }

  // --- Analysis ---
  const current   = prices[prices.length - 1];
  const predPrices = prediction.map(p => p.price);
  const minPred   = Math.min(...predPrices);
  const maxPred   = Math.max(...predPrices);
  const minIdx    = predPrices.indexOf(minPred);
  const bestDay   = prediction[minIdx].date;
  const lastPred  = predPrices[6];
  const pctChange = ((lastPred - current) / current) * 100;
  const savings   = current - minPred;

  // Drop detection: any single day drop > 5%
  const dropDetected = predPrices.some((p, i) =>
    i > 0 && (p - predPrices[i - 1]) / predPrices[i - 1] * 100 < -5
  );

  // Confidence: R² × data factor (need 10 points for full confidence)
  const confidence = Math.round(Math.min(100, r2 * Math.min(1, n / 10) * 100) * 10) / 10;

  // Trend & insight
  let trend, insight, recommendation;
  if (pctChange < -3) {
    trend          = 'falling';
    recommendation = 'WAIT';
    insight = dropDetected
      ? `🚨 Sharp price drop predicted! Price expected to fall by ₹${Math.abs(savings).toFixed(0)} in ${minIdx + 1} day(s). Best time to buy: ${bestDay}.`
      : `📉 Price is trending down ${Math.abs(pctChange).toFixed(1)}% over next 7 days. Best time to buy: ${bestDay} at ₹${Math.round(minPred).toLocaleString('en-IN')}.`;
  } else if (pctChange > 3) {
    trend          = 'rising';
    recommendation = 'BUY NOW';
    insight        = `📈 Price likely to rise ${Math.abs(pctChange).toFixed(1)}% over 7 days. Buy now — current price ₹${Math.round(current).toLocaleString('en-IN')} is the best available.`;
  } else {
    trend          = 'stable';
    recommendation = 'MONITOR';
    insight        = `➡️ Price is stable (±${Math.abs(pctChange).toFixed(1)}%). No rush to buy. Lowest predicted: ₹${Math.round(minPred).toLocaleString('en-IN')} on ${bestDay}.`;
  }
  if (confidence < 40)
    insight += ` ⚠️ Low confidence — update price more often for better accuracy.`;

  // History for graph (last 14 entries)
  const historyGraph = history.slice(-14).map(r => ({
    date:  new Date(r.recorded_at).toISOString().split('T')[0],
    price: parseFloat(r.price),
  }));

  return {
    product_id:          pid,
    product_name:        name,
    product_category:    category,
    current_price:       current,
    data_points:         n,
    r2_score:            Math.round(r2 * 10000) / 10000,
    prediction,
    history:             historyGraph,
    insight,
    trend,
    recommendation,
    confidence,
    drop_detected:       dropDetected,
    best_buy_day:        bestDay,
    min_predicted:       Math.round(minPred * 100) / 100,
    max_predicted:       Math.round(maxPred * 100) / 100,
    expected_savings:    Math.max(0, Math.round(savings * 100) / 100),
    expected_change_pct: Math.round(pctChange * 100) / 100,
  };
}

module.exports = router;